### PR TITLE
Invoice Calculator Code Optimizations and New Tests

### PR DESCRIPTION
**Pull request by reshmabidikar**
- Includes `InvoiceCalculatorUtils` optimizations
- Added additional tests to [TestDefaultInvoiceUserApi ](https://github.com/killbill/killbill/blob/1816bb273b79a665ecde628516bf6812dc773ad0/invoice/src/test/java/org/killbill/billing/invoice/api/user/TestDefaultInvoiceUserApi.java) (Not sure if this is the right place but added a few tests that would test the changes done to `InvoiceCalculatorUtils` )

### CODE REVIEW

**Review by sbrossie — COMMENTED** [review]
Looks very good - most of my comments are small Nit comments that are very arguable, mostly stylistic.

File: invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java [hunk 1]
```diff
@@ -35,14 +36,47 @@
...
-    public static boolean isInvoiceAdjustmentItem(final InvoiceItem invoiceItem, final Iterable<InvoiceItem> otherInvoiceItems) {
+    public static boolean isInvoiceAdjustmentItem(final InvoiceItem invoiceItemToCheck, final Iterable<InvoiceItem> invoiceItems) {
         // Invoice level credit, i.e. credit adj, but NOT on its on own invoice
-        return (InvoiceItemType.CREDIT_ADJ.equals(invoiceItem.getInvoiceItemType()) &&
-                !(Iterables.size(otherInvoiceItems) == 1 &&
-                  InvoiceItemType.CBA_ADJ.equals(otherInvoiceItems.iterator().next().getInvoiceItemType()) &&
-                  otherInvoiceItems.iterator().next().getInvoiceId().equals(invoiceItem.getInvoiceId()) &&
-                  otherInvoiceItems.iterator().next().getAmount().compareTo(invoiceItem.getAmount().negate()) == 0));
+
+        if ((InvoiceItemType.CREDIT_ADJ.equals(invoiceItemToCheck.getInvoiceItemType()) && (Iterables.size(invoiceItems) != 2 || !isCreditInvoice(invoiceItems)))) {
```
[the remark this answers is not in the document]
Comment (reshmabidikar) [reply invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java]: Fixed

File: invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java [hunk 2]
```diff
@@ -35,14 +36,47 @@
...
+            return true;
+        } else {
+            return false;
+        }
+
+    }
+
+    private static boolean isCreditInvoice(final Iterable<InvoiceItem> invoiceItems) { //checks if the invoiceItems correspond to a CREDIT, that is there are only 2 items out of which one is a CREDIT_ADJ and the other is a CBA_ADJ
+
+        if (Iterables.size(invoiceItems) != 2) { //if the number of invoice items are not 2, the invoice DOES NOT correspond to a CREDIT, so return false
```
[the remark this answers is not in the document]
Comment (reshmabidikar) [reply invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java]: Fixed, removed all comments, they were mostly for me to know what I was doing :)

File: invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java [hunk 3]
```diff
@@ -35,14 +36,47 @@
...
+            return false;
+        }
+
+        Iterator<InvoiceItem> itr = invoiceItems.iterator();
```
[the remark this answers is not in the document]
Comment (reshmabidikar) [reply invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java]: Fixed

File: invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java [hunk 4]
```diff
@@ -35,14 +36,47 @@
...
+
+        InvoiceItem item1 = itr.next();
+        InvoiceItem item2 = itr.next();
+
+        //If there is no item type of CREDIT_ADJ, return false
```
[the remark this answers is not in the document]
Comment (reshmabidikar) [reply invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java]: Fixed

File: invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java [hunk 5]
```diff
@@ -35,14 +36,47 @@
...
+        if (!InvoiceItemType.CREDIT_ADJ.equals(item1.getInvoiceItemType()) && !InvoiceItemType.CREDIT_ADJ.equals(item2.getInvoiceItemType())) {
+            return false;
+        }
+
+        //If item1 is CREDIT_ADJ and item2 is CBA_ADJ or vice versa, return true
+        if ((InvoiceItemType.CREDIT_ADJ.equals(item1.getInvoiceItemType()) && isCbaAdjItemInCreditInvoice(item1, item2)) || (InvoiceItemType.CREDIT_ADJ.equals(item2.getInvoiceItemType()) && isCbaAdjItemInCreditInvoice(item2, item1))) {
```
[the remark this answers is not in the document]
Comment (reshmabidikar) [reply invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java]: Fixed

File: invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java [hunk 6]
```diff
@@ -35,14 +36,47 @@
...
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private static boolean isCbaAdjItemInCreditInvoice(final InvoiceItem creditAdjItem, final InvoiceItem itemToCheck) { // check if it is a CBA_ADJ item within a CREDIT invoice of its own
+        if (InvoiceItemType.CBA_ADJ.equals(itemToCheck.getInvoiceItemType()) && itemToCheck.getInvoiceId().equals(creditAdjItem.getInvoiceId()) && itemToCheck.getAmount().compareTo(creditAdjItem.getAmount().negate()) == 0) {
```
[the remark this answers is not in the document]
Comment (reshmabidikar) [reply invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java]: Fixed

File: invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java [hunk 7]
```diff
@@ -111,29 +144,26 @@ public boolean apply(final InvoiceItem input) {
...
+
+            Iterator<InvoiceItem> itr = invoiceItems.iterator();
+
+            InvoiceItem item1 = itr.next();
+            InvoiceItem item2 = itr.next();
+
+            if (InvoiceItemType.CREDIT_ADJ.equals(item1.getInvoiceItemType())) {
+                amountAdjusted = amountAdjusted.add(item1.getAmount());
+            } else {
+                amountAdjusted = amountAdjusted.add(item2.getAmount());
```
[the remark this answers is not in the document]
Comment (reshmabidikar) [reply invoice/src/main/java/org/killbill/billing/invoice/calculator/InvoiceCalculatorUtils.java:154]: Fixed

File: invoice/src/test/java/org/killbill/billing/invoice/api/user/TestDefaultInvoiceUserApi.java [hunk 8]
```diff
@@ -619,11 +610,65 @@ public boolean apply(final InvoiceItem invoiceItem) {
...
+    public void testCreditOnDraftInvoice() throws Exception {
+
+        final Account account = invoiceUtil.createAccount(callContext);
+        final UUID accountId = account.getId();
+
+        // Create external charge
+        final BigDecimal externalChargeAmount = BigDecimal.TEN;
+        final InvoiceItem externalCharge = new ExternalChargeInvoiceItem(null, accountId, null, UUID.randomUUID().toString(), clock.getUTCToday(), null, externalChargeAmount, accountCurrency, null);
+
+        //create invoice in draft status with external charge
```
[the remark this answers is not in the document]
Comment (reshmabidikar) [reply invoice/src/test/java/org/killbill/billing/invoice/api/user/TestDefaultInvoiceUserApi.java]: Fixed

File: invoice/src/test/java/org/killbill/billing/invoice/api/user/TestDefaultInvoiceUserApi.java [hunk 9]
```diff
@@ -619,11 +610,65 @@ public boolean apply(final InvoiceItem invoiceItem) {
...
+
+        //Post a credit on draft invoice
+        final BigDecimal creditAmount = new BigDecimal(3);
+        final InvoiceItem credit = new CreditAdjInvoiceItem(externalChargeInvoice.getId(), accountId, clock.getUTCToday(), "Adding credit", creditAmount, accountCurrency, null);
+        invoiceUserApi.insertCredits(accountId, clock.getUTCToday(), ImmutableList.<InvoiceItem>of(credit), true, null, callContext);
+
+        final Invoice updatedExternalChargeInvoice = invoiceUserApi.getInvoice(externalChargeInvoice.getId(), callContext);
+        Assert.assertEquals(updatedExternalChargeInvoice.getChargedAmount().compareTo(new BigDecimal(7)), 0); // credit is adjusted in the amountCharged since invoice is in DRAFT status //TOOD: how is this different from the test below?
+        Assert.assertEquals(updatedExternalChargeInvoice.getBalance().compareTo(BigDecimal.ZERO), 0);// invoice is in DRAFT status, so balance is 0
+    }
```
[the remark this answers is not in the document]
Comment (reshmabidikar) [reply invoice/src/test/java/org/killbill/billing/invoice/api/user/TestDefaultInvoiceUserApi.java:649]: Fixed

**Review by sbrossie — APPROVED** [review]

**Pull request merged**